ethereum-contracts/contracts/Peggy.sol

```diff
@@ -1,95 +1,266 @@
     /*
-     * @param to          bytes representation of destination address
-     * @param value       value of transference
-     * @param token       token address in origin chain (0x0 if Ethereum, Cosmos for other values)
-     * @param chain       bytes respresentation of the destination chain (not used in MVP, for incentivization of relayers)
-     * @param signers     indexes of each validator
-     * @param v           array of recoverys id
-     * @param r           array of outputs of ECDSA signature
-     * @param s           array of outputs of ECDSA signature
+     * @dev: Unlocks ethereum/erc20 tokens, called by relayer.
+     *
+     *       This is a shortcut utility method for testing purposes.
+     *       In the future bidirectional system, unlocking functionality
+     *       will be guarded by validator signatures.
+     *
+     * @param _id: Unique key of the item.
      */
-    function unlock(address to, address token, uint64 amount, uint[] signers, uint8[] v, bytes32[] r, bytes32[] s) external returns (bool) {
-        bytes32 hashData = keccak256(to, token, amount);
-        require(Valset.verifyValidators(hashData, signers, v, r, s));
-        if (token == address(0)) {
-          to.transfer(amount);
-        } else if (cosmosTokenAddresses[token]) {
-          CosmosERC20(token).mint(to, amount);
-        } else {
-          require(ERC20(token).transfer(to, amount));
-        }
-        Unlock(to, token, amount);
+    function unlock(
+        bytes32 _id
+    )
+        onlyRelayer
+        canDeliver(_id)
+        external
+        returns (bool)
+    {
+        require(isLocked(_id));
+
+        // Transfer item's funds and unlock it
+        (address payable sender,
+            address token,
+            uint256 amount,
+            uint256 uniqueNonce) = complete(_id);
+
+        //Emit unlock event
+        emit LogUnlock(
+            _id,
+            sender,
+            token,
+            amount,
+            uniqueNonce
+        );
         return true;
     }
 
-    function newCosmosERC20(string name, uint decimals, uint[] signers, uint8[] v, bytes32[] r, bytes32[] s) external returns (address addr) {
-        require(cosmosTokens[name] == address(0));
+    /*
+     * @dev: Withdraws ethereum/erc20 tokens, called original sender.
+     *
+     *       This is a backdoor utility method included for testing,
+     *       purposes, allowing users to withdraw their funds. This
+     *       functionality will be removed in production.
+     *
+     * @param _id: Unique key of the item.
+     */
+    function withdraw(
+        bytes32 _id
+    )
+        onlySender(_id)
+        canDeliver(_id)
+        external
+        returns (bool)
+    {
+        require(isLocked(_id));
 
-        bytes32 hashData = keccak256(name, decimals);
-        require(Valset.verifyValidators(hashData, signers, v, r, s));
+        // Transfer item's funds and unlock it
```

Comment:
what's an item in this context ?

Reply:
An item refers to a single incoming transaction containing funds to be locked. Originally named 'escrow', the terminology was updated to 'item' to prevent confusion regarding the fact that the funds are not technically held in escrow for the unidirectional implementation. Happy to update the naming conventions again, should there be consensus about a more suitable name for an instance of locked funds.

Reply:
I see. We use the term `Msg` on the SDK (See [`sdk.Msg`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go) and [`auth.StdTx`](https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/stdtx.go#L22) for reference). I think that's a more accurate term for this case

Reply:
Items aren't 1:1 with Msgs, as Items serve as storage for the original sender's funds and include a unique id which isn't currently reflected in the Msg sent on Cosmos - but I agree that 'Msg' is a much more accurate name than 'Item'. @cwgoes what are your thoughts on the naming conventions?